src/V3Const.cpp

```diff
@@ -590,6 +590,86 @@ class ConstVisitor : public AstNVisitor {
+        if (const AstConst* const np = VN_CAST_CONST(nodep, Const)) {
+            out.opAssign(np->num());
+        } else if (const AstOr* const np = VN_CAST_CONST(nodep, Or)) {
+            // @wsnyder Do I need to check operands are sized and have the same width?
+            V3Number lb(nodep, np->lhsp()->widthMin());
+            V3Number rb(nodep, np->rhsp()->widthMin());
+            setBits(lb, np->lhsp());
+            setBits(rb, np->rhsp());
+            out.opOr(lb, rb);
+        } else {
+            // Pretend we don't know anything
+            out.setAllBits0();
+        }
+    }
+
+    // 'out' will have those bits set, which are known to be clear
+    void clrBits(V3Number& out, AstNode* nodep) {
+        // @wsnyder Do I need some sizedness test here, or are they already well formed?
+        if (const AstConst* const np = VN_CAST_CONST(nodep, Const)) {
+            out.opNot(np->num());
+        } else if (const AstOr* const np = VN_CAST_CONST(nodep, Or)) {
+            // @wsnyder Do I need to check operands are sized and have the same width?
+            V3Number lb(nodep, np->lhsp()->widthMin());
+            V3Number rb(nodep, np->rhsp()->widthMin());
+            clrBits(lb, np->lhsp());
+            clrBits(rb, np->rhsp());
+            out.opAnd(lb, rb);
+        } else if (const AstVarRef* const np = VN_CAST_CONST(nodep, VarRef)) {
+            out.setAllBits0();
+            for (int bit = np->varp()->widthMin(); bit < nodep->widthMin(); ++bit) {
+                out.setBit(bit, 1);
+            }
+        } else {
+            // Pretend we don't know anything
+            out.setAllBits0();
+        }
+    }
+
+    bool checkKnownBits(AstNodeBiop* nodep) {
+        // Check if the sides have enough known bits to to determine the result
+        // This mostly helps to squash compiler warnings so we are not doing it
+        // too excessively.
+        const int width = nodep->lhsp()->widthMin();
+        // @wsnyder Is this check necessary, or are they already well formed?
+        if (width == 0 || width != nodep->lhsp()->widthMin()) return false;
+
+        V3Number lb(nodep, width);
+        V3Number rb(nodep, width);
+        V3Number tst(nodep, width);
+
+        // Compare bits clear in LHS with bits set in RHS
+        clrBits(lb, nodep->lhsp());
+        setBits(rb, nodep->rhsp());
+        if (!tst.opAnd(lb, rb).isEqZero()) {
+            if (VN_IS(nodep, Eq)) {
+                replaceZero(nodep);
+            } else {  // Neq
+                replaceNum(nodep, 1);
+            }
+            return true;
+        }
+
+        // Compare bits set in LHS with bits clear in RHS
+        setBits(lb, nodep->lhsp());
+        clrBits(rb, nodep->rhsp());
+        if (!tst.opAnd(lb, rb).isEqZero()) {
+            if (VN_IS(nodep, Eq)) {
+                replaceZero(nodep);
+            } else {  // Neq
+                replaceNum(nodep, 1);
+            }
+            return true;
+        }
```

Comment:
Maybe save some lines.

```
        // Compare bits clear in LHS with bits set in RHS
        clrBits(lbc, nodep->lhsp());
        setBits(rbs, nodep->rhsp());
        setBits(lbs, nodep->lhsp());
        clrBits(rbc, nodep->rhsp());
        if (!tst.opAnd(lbs, rbc).isEqZero()
            || !tst.opAnd(lbc, rbs).isEqZero()) {
            if (VN_IS(nodep, Eq)) {
```

Reply:
Yeah, I'm torn about that one as the current form short circuits earlier. On the other hand that very rarely happens, and I could short circut even earalier if one of the vector is already all zeros... Probably will go with your version for simplicity until it shows up in a profile, which I doubt it will.